### ISSUE TITLE
Detect webnavigation changes

### ISSUE THREAD

**Issue by Revadike** [tfedor/AugmentedSteam#375]
So, Steam has some web pages that can load dynamically. This is problematic, as this will remove any augmentation done by the extension. It has to be redone.
To also change the URL, they use the `history.pushstate` function. Sadly, modern web browsers have no event for this. Chrome extensions however, can make use of `webNavigation` to detect this event, by attaching a listener to `chrome.webNavigation.onHistoryStateUpdated`. It does require you to add additional `webNavigation` permission.
I'd add this feature myself, but I am not entirely sure how you would this to be implemented. Also, this listener has to be done in a background script. So for other scripts to communicate we probably have to use some messages system.

Here is an example:
1. Go to https://steamcommunity.com/my/friends
2. Observe the friend sorting augmentation that was added by Augmented Steam.
3. Now load another page by clicking e.g. "Add a friend"
4. Now, go back to "Your friends"
5. Observe how the friend sorting augmentation is now missing

**Comment by MxtOUT** [comment]
Just a note, we already have a messaging system, look at the `Background` class

**Comment by Revadike** [comment]
> Just a note, we already have a messaging system, look at the `Background` class

Yes, I saw, but this is only used to send messages to background and no from background to e.g. community.js, which is needed as the event handler needs to be in the background.

**Comment by Revadike** [comment]
![image](https://user-images.githubusercontent.com/4411977/62786196-699f9a80-bac2-11e9-81ce-383a045acb79.png)
This is an example response from the API event.

**Comment by Revadike** [comment]
Not all dynamically loaded content is loaded by updating navigation. Take the steam group tabs for example.
I've concluded, the only way to do this for all cases, is to either proxy a native steam functions that get called during this event, like this:
```js
ExtensionLayer.runInPageContext(function() {
            let originalFn = OnContentLoaded;
            OnContentLoaded = (...arguments) => {
                originalFn(...arguments);
                Messenger.postMessage("newContent");
            };
        });

        Messenger.addMessageListener("newContent", () => {
            this.invokedFn();
        }, true);
```

or observe ajax requests.
Both have their up- and downsides.
I've loaded into observing `var`s in the window context, but that is impossible. Only configurable properties of an object are watchable, which is not the case with `var`s.

**Comment by Revadike** [comment]
Another thing to look out for is to not apply mutation more than once. So, when our function is called we have to put in a guard to check if it's not already applied.

**Comment by Revadike** [comment]
Actually, since steam doesn't seem to use `var`/`let`/`const` for their global variable, observing them is totally possible!

```js
let proxy = g_sessionID.valueOf();
delete g_sessionID;
Object.defineProperty(this, 'g_sessionID', {
  get: function () { return proxy; },
  set: function (v) {
    proxy = v;
    console.log('Value changed! New value: ' + v);
  }
});
```

**Comment by Revadike** [comment]
`Ajax.Responders.register({ onComplete: ourHandler })` doesn't work for every case. In particular, where they use `CAjaxSubPageController` like in https://steamcommunity.com/my/friends/. In this case, we can use `$J.ajaxSetup({ complete: ourHandler })` instead, since that's how they perform ajax requests:
![image](https://user-images.githubusercontent.com/4411977/62884898-df606c00-bd37-11e9-8c15-9aceab7388f1.png)

**Comment by Revadike** [comment]
I think I am ready to write a global handler/class for this issue, using the above method. However, since some of my pull requests can benefit from this, it'd be annoying to have them all open and not merged.

**Comment by Revadike** [comment]
In short, my idea is to hook our handler to steam's ajax requests. In our handler we will either use a whitelist of request filters, or always just rerun the store.js/community.js init function. It will also require every function to have guards to check if the function isn't already present/executed on the (new) page.

**Comment by candela97** [comment]
How about `MutationObserver`? Something like this works for Groups and Friends pages:
````
    let node = document.getElementById("subpage_container");
    if (node) {
        new MutationObserver(mutations => {
            for (let mutation of mutations) {
                for (let node of mutation.addedNodes) {
                    if (node.id === "friends_list") {
                        new FriendsPageClass();
                    } else if (node.id === "groups_list") {
                        new GroupsPageClass();
                    }
                }
            }
        }).observe(node, { childList: true });
    }
````

**Comment by Revadike** [comment]
> How about `MutationObserver`? Something like this works for Groups and Friends pages:
> 
> 
> ```
>     let node = document.getElementById("subpage_container");
>     if (node) {
>         new MutationObserver(mutations => {
>             for (let mutation of mutations) {
>                 for (let node of mutation.addedNodes) {
>                     if (node.id === "friends_list") {
>                         new FriendsPageClass();
>                     } else if (node.id === "groups_list") {
>                         new GroupsPageClass();
>                     }
>                 }
>             }
>         }).observe(node, { childList: true });
>     }
> ```

Sure, but a general solution would be nicer. This solution requires manual intervention for each dynamic content change.

**Comment by candela97** [comment]
Hmm, maybe it can be a temporary solution? So far I only noticed issues on those two pages, and there aren't any feature requests for other related pages at the moment.

**Comment by MxtOUT** [comment]
The first step is done, but I assume this isn't the end.
Reopening

**Comment by candela97** [comment]
Another area where this comes up is on groups page: all subpages apart from discussions are dynamically loaded. But there shouldn't be any issues atm because there're no subpage-specific features.
